src/components/network/layers/arrow-layer-vertex.glsl.js

```diff
@@ -0,0 +1,134 @@
+in float instanceArrowDistance;
+in vec4 instanceColor;
+in float instanceSpeedFactor;
+in int instanceLinePositionsTextureOffset;
+in int instanceLineDistancesTextureOffset;
+in int instanceLinePointCount;
+in float instanceLineDistance;
+in float instanceArrowDirection;
+
+uniform float sizeMinPixels;
+uniform float sizeMaxPixels;
+uniform float timestamp;
+uniform sampler2D linePositionsTexture;
+uniform sampler2D lineDistancesTexture;
+uniform float maxTextureSize;
+
+out vec4 vFillColor;
+out float shouldDiscard;
+
+/**
+ * Calculate 2 dimensions texture index from flat index. 
+ */
+ivec2 calulateTextureIndex(int flatIndex) {
+  int x = int(mod(float(flatIndex), maxTextureSize));
+  int y = int(trunc(float(flatIndex) / maxTextureSize));
+  return ivec2(x, y);
+}
+
+/**
+ * Fetch WGS84 position from texture for a given point of the line.  
+ */
+vec3 fetchLinePosition(int point) {
+  int flatIndex = instanceLinePositionsTextureOffset + point;
+  ivec2 textureIndex = calulateTextureIndex(flatIndex); 
+  return vec3(texelFetch(linePositionsTexture, textureIndex, 0).xy, 0);
+}
+
+/**
+ * Fetch distance (in meters from the start of the line) from texture for a point of the line.  
+ */
+float fetchLineDistance(int point) {
+  int flatIndex = instanceLineDistancesTextureOffset + point;
+  ivec2 textureIndex = calulateTextureIndex(flatIndex);
+  return texelFetch(lineDistancesTexture, textureIndex, 0).x;
+}
+
+/**
+ * Find the first point of the line that is after a given distance from the start (first line point).
+ * (implemented using a binary search)
+ */
+int findFirstLinePointAfterDistance(float distance) {
+  int firstPoint = 1;
+  int lastPoint = instanceLinePointCount - 1;
+  if (firstPoint == lastPoint) {
+      return firstPoint;
+  }
+  // variable length loops are not supported in GLSL, instanceLinePointCount is an upper bound that
+  // will never be reached as binary search complexity is in O(log(instanceLinePointCount))
+  for (int i = 0; i < instanceLinePointCount; i++) {
+      if (firstPoint + 1 == lastPoint) {
+          float firstPointDistance = fetchLineDistance(firstPoint);
+          if (firstPointDistance > distance) {
+              return firstPoint;
+          } else {
+              return lastPoint;
+          }
+      }
+      int middlePoint = int(ceil(float(firstPoint + lastPoint) / 2.0));
+      float middlePointDistance = fetchLineDistance(middlePoint);      
+      if (middlePointDistance < distance) {
+         firstPoint = middlePoint;
+      } else {
+         lastPoint = middlePoint;
+      }
+  }
+}
+
+void main(void) {
+  if (instanceArrowDirection < 1.0) {
+      vFillColor = vec4(0, 0, 0, 0);
+      shouldDiscard = 1.0;
+  } else {
+      // arrow distance from the line start shifted with current timestamp
+      float arrowDistance = mod(instanceLineDistance * instanceArrowDistance + (instanceArrowDirection < 2.0 ? 1.0 : -1.0) * timestamp * instanceSpeedFactor, instanceLineDistance);
```

Comment:
float arrowDistance = mod((instanceArrowDirection < 2.0 ? 1.0 : -1.0) * timestamp * instanceSpeedFactor, instanceLineDistance);

mod(a*b+X, a) == mod(X, a) no ??

Reply:
even if b < 1 ?

Reply:
Ah yeah only when b is an integer of course. So you're right, here it's absolutely needed other all arrows are at the same position

Reply:
somewhere we should add a comment :
instanceArrowDistance: a float in interval [0,1] describing the initial position of the arrow

Reply:
comment added